// deep-iterator@1.1.0/lib/deep-iterator.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

exports.default = deepIterator;

var _search = require('./search');

var search = _interopRequireWildcard(_search);

var _rootNode = require('./root-node');

var _rootNode2 = _interopRequireDefault(_rootNode);

var _seen = require('./seen');

var _seen2 = _interopRequireDefault(_seen);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _marked = [deepIterator].map(_regenerator2.default.mark);

function deepIterator(rootElement) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var config, node, seen;
  return _regenerator2.default.wrap(function deepIterator$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          config = {
            onlyLeaves: false,
            circularReference: 'leaf',
            search: 'dfsPreOrder',
            iterateOverObject: true,
            skipIteration: function skipIteration() {
              return false;
            }
          };


          if (options.onlyLeaves !== undefined) {
            config.onlyLeaves = options.onlyLeaves;
          }
          if (options.circularReference !== undefined) {
            config.circularReference = options.circularReference;
          }
          if (options.iterateOverObject !== undefined) {
            config.iterateOverObject = options.iterateOverObject;
          }
          if (options.skipIteration !== undefined) {
            config.skipIteration = options.skipIteration;
          }

          if (!(options.search !== undefined)) {
            _context.next = 9;
            break;
          }

          if (options.search in search) {
            _context.next = 8;
            break;
          }

          throw new Error('The search algorithm ' + options.search + ' is incorrect.');

        case 8:
          config.search = options.search;

        case 9:
          node = new _rootNode2.default(rootElement, config);
          seen = (0, _seen2.default)(config.circularReference);
          return _context.delegateYield(search[config.search](node, config.onlyLeaves, seen), 't0', 12);

        case 12:
        case 'end':
          return _context.stop();
      }
    }
  }, _marked[0], this);
}
