// deep-iterator@1.1.0/lib/node.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _generators = require('./generators');

var _tag = require('./tag');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Node = function () {
  function Node(key, value, parentNode, isCircular) {
    (0, _classCallCheck3.default)(this, Node);

    this._value = value;
    this._key = key;
    this._parent = parentNode.value, this._parentNode = parentNode;
    this._selectGenerator = parentNode._selectGenerator;
    this._tag = (0, _tag.getTag)(value);
    this._generator = this._selectGenerator.getGenerator(this._tag);
    this._isCircular = isCircular;
    this._isLeaf = this._generator === _generators.LEAF || this._isCircular;
    if (!this._isLeaf && this._selectGenerator.skipIteration(this)) {
      this._isLeaf = true;
      this._generator = _generators.LEAF;
    }
  }

  (0, _createClass3.default)(Node, [{
    key: 'createIterator',
    value: function createIterator() {
      if (this._generator === _generators.LEAF) {
        throw new Error("createIterator called on a non iterable node.");
      }
      return this._generator(this.value);
    }
  }, {
    key: 'isCircular',
    value: function isCircular() {
      return this._isCircular;
    }
  }, {
    key: 'isLeaf',
    value: function isLeaf() {
      return this._isLeaf;
    }
  }, {
    key: 'isNull',
    value: function isNull() {
      return this.type === 'Null';
    }
  }, {
    key: 'isUndefined',
    value: function isUndefined() {
      return this.type === 'Undefined';
    }
  }, {
    key: 'isBoolean',
    value: function isBoolean() {
      return this.type === 'Boolean';
    }
  }, {
    key: 'isNumber',
    value: function isNumber() {
      return this.type === 'Number';
    }
  }, {
    key: 'isString',
    value: function isString() {
      return this.type === 'String';
    }
  }, {
    key: 'isSymbol',
    value: function isSymbol() {
      return this.type === 'Symbol';
    }
  }, {
    key: 'isDate',
    value: function isDate() {
      return this.type === 'Date';
    }
  }, {
    key: 'isRegExp',
    value: function isRegExp() {
      return this.type === 'RegExp';
    }
  }, {
    key: 'isFunction',
    value: function isFunction() {
      return this.type === 'Function';
    }
  }, {
    key: 'isGeneratorFunction',
    value: function isGeneratorFunction() {
      return this.type === 'GeneratorFunction';
    }
  }, {
    key: 'isPromise',
    value: function isPromise() {
      return this.type === 'Promise';
    }
  }, {
    key: 'isArray',
    value: function isArray() {
      return this.type === 'Array';
    }
  }, {
    key: 'isSet',
    value: function isSet() {
      return this.type === 'Set';
    }
  }, {
    key: 'isMap',
    value: function isMap() {
      return this.type === 'Map';
    }
  }, {
    key: 'isUserDefinedIterable',
    value: function isUserDefinedIterable() {
      return this.type === 'UserDefinedIterable';
    }
  }, {
    key: 'isNonIterableObject',
    value: function isNonIterableObject() {
      return this.type === 'NonIterableObject';
    }
  }, {
    key: 'value',
    get: function get() {
      return this._value;
    }
  }, {
    key: 'key',
    get: function get() {
      return this._key;
    }
  }, {
    key: 'parent',
    get: function get() {
      return this._parent;
    }
  }, {
    key: 'parentNode',
    get: function get() {
      return this._parentNode;
    }
  }, {
    key: 'path',
    get: function get() {
      // evaluated only when needed - could memoize it
      return [].concat((0, _toConsumableArray3.default)(this.parentNode.path), [this.key]);
    }
  }, {
    key: 'type',
    get: function get() {
      // simple memoization
      /* istanbul ignore next */
      if (this._type === undefined) {
        this._type = (0, _tag.getTypeFromTag)(this._tag);
      }
      return this._type;
    }
  }]);
  return Node;
}();

exports.default = Node;
