// deep-iterator@1.1.0/lib/search.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

exports.dfsPreOrder = dfsPreOrder;
exports.dfsPostOrder = dfsPostOrder;
exports.bfs = bfs;

var _node = require('./node');

var _node2 = _interopRequireDefault(_node);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _marked = [dfsPreOrder, dfsPostOrder, bfs].map(_regenerator2.default.mark);

function dfsPreOrder(node, onlyLeaves, parentSeen) {
  var seen, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, _step$value, key, value;

  return _regenerator2.default.wrap(function dfsPreOrder$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          if (!node.isLeaf()) {
            _context.next = 4;
            break;
          }

          _context.next = 3;
          return node;

        case 3:
          return _context.abrupt('return');

        case 4:
          if (onlyLeaves) {
            _context.next = 7;
            break;
          }

          _context.next = 7;
          return node;

        case 7:
          seen = parentSeen.add(node.value);
          _iteratorNormalCompletion = true;
          _didIteratorError = false;
          _iteratorError = undefined;
          _context.prev = 11;
          _iterator = (0, _getIterator3.default)(node.createIterator());

        case 13:
          if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
            _context.next = 19;
            break;
          }

          _step$value = (0, _slicedToArray3.default)(_step.value, 2), key = _step$value[0], value = _step$value[1];
          return _context.delegateYield(dfsPreOrder(new _node2.default(key, value, node, seen.has(value)), onlyLeaves, seen), 't0', 16);

        case 16:
          _iteratorNormalCompletion = true;
          _context.next = 13;
          break;

        case 19:
          _context.next = 25;
          break;

        case 21:
          _context.prev = 21;
          _context.t1 = _context['catch'](11);
          _didIteratorError = true;
          _iteratorError = _context.t1;

        case 25:
          _context.prev = 25;
          _context.prev = 26;

          if (!_iteratorNormalCompletion && _iterator.return) {
            _iterator.return();
          }

        case 28:
          _context.prev = 28;

          if (!_didIteratorError) {
            _context.next = 31;
            break;
          }

          throw _iteratorError;

        case 31:
          return _context.finish(28);

        case 32:
          return _context.finish(25);

        case 33:
        case 'end':
          return _context.stop();
      }
    }
  }, _marked[0], this, [[11, 21, 25, 33], [26,, 28, 32]]);
}

function dfsPostOrder(node, onlyLeaves, parentSeen) {
  var seen, _iteratorNormalCompletion2, _didIteratorError2, _iteratorError2, _iterator2, _step2, _step2$value, _key, _value;

  return _regenerator2.default.wrap(function dfsPostOrder$(_context2) {
    while (1) {
      switch (_context2.prev = _context2.next) {
        case 0:
          if (!node.isLeaf()) {
            _context2.next = 4;
            break;
          }

          _context2.next = 3;
          return node;

        case 3:
          return _context2.abrupt('return');

        case 4:
          seen = parentSeen.add(node.value);
          _iteratorNormalCompletion2 = true;
          _didIteratorError2 = false;
          _iteratorError2 = undefined;
          _context2.prev = 8;
          _iterator2 = (0, _getIterator3.default)(node.createIterator());

        case 10:
          if (_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done) {
            _context2.next = 16;
            break;
          }

          _step2$value = (0, _slicedToArray3.default)(_step2.value, 2), _key = _step2$value[0], _value = _step2$value[1];
          return _context2.delegateYield(dfsPostOrder(new _node2.default(_key, _value, node, seen.has(_value)), onlyLeaves, seen), 't0', 13);

        case 13:
          _iteratorNormalCompletion2 = true;
          _context2.next = 10;
          break;

        case 16:
          _context2.next = 22;
          break;

        case 18:
          _context2.prev = 18;
          _context2.t1 = _context2['catch'](8);
          _didIteratorError2 = true;
          _iteratorError2 = _context2.t1;

        case 22:
          _context2.prev = 22;
          _context2.prev = 23;

          if (!_iteratorNormalCompletion2 && _iterator2.return) {
            _iterator2.return();
          }

        case 25:
          _context2.prev = 25;

          if (!_didIteratorError2) {
            _context2.next = 28;
            break;
          }

          throw _iteratorError2;

        case 28:
          return _context2.finish(25);

        case 29:
          return _context2.finish(22);

        case 30:
          if (onlyLeaves) {
            _context2.next = 33;
            break;
          }

          _context2.next = 33;
          return node;

        case 33:
        case 'end':
          return _context2.stop();
      }
    }
  }, _marked[1], this, [[8, 18, 22, 30], [23,, 25, 29]]);
}

// transpiled generator has uncovered branches
// istanbul issue #645
/* istanbul ignore next */
function bfs(rootNode, onlyLeaves, parentSeen) {
  var queue, i, node, seen, _iteratorNormalCompletion3, _didIteratorError3, _iteratorError3, _iterator3, _step3, _step3$value, _key2, _value2;

  return _regenerator2.default.wrap(function bfs$(_context3) {
    while (1) {
      switch (_context3.prev = _context3.next) {
        case 0:
          queue = [{ node: rootNode, seen: parentSeen }];
          i = 0;

        case 2:
          if (!(i < queue.length)) {
            _context3.next = 35;
            break;
          }

          node = queue[i].node;

          if (!node.isLeaf()) {
            _context3.next = 9;
            break;
          }

          _context3.next = 7;
          return node;

        case 7:
          _context3.next = 32;
          break;

        case 9:
          if (onlyLeaves) {
            _context3.next = 12;
            break;
          }

          _context3.next = 12;
          return node;

        case 12:
          seen = queue[i].seen.add(node.value);
          _iteratorNormalCompletion3 = true;
          _didIteratorError3 = false;
          _iteratorError3 = undefined;
          _context3.prev = 16;

          for (_iterator3 = (0, _getIterator3.default)(node.createIterator()); !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
            _step3$value = (0, _slicedToArray3.default)(_step3.value, 2), _key2 = _step3$value[0], _value2 = _step3$value[1];

            queue.push({
              node: new _node2.default(_key2, _value2, node, seen.has(_value2)),
              seen: seen
            });
          }
          _context3.next = 24;
          break;

        case 20:
          _context3.prev = 20;
          _context3.t0 = _context3['catch'](16);
          _didIteratorError3 = true;
          _iteratorError3 = _context3.t0;

        case 24:
          _context3.prev = 24;
          _context3.prev = 25;

          if (!_iteratorNormalCompletion3 && _iterator3.return) {
            _iterator3.return();
          }

        case 27:
          _context3.prev = 27;

          if (!_didIteratorError3) {
            _context3.next = 30;
            break;
          }

          throw _iteratorError3;

        case 30:
          return _context3.finish(27);

        case 31:
          return _context3.finish(24);

        case 32:
          i++;
          _context3.next = 2;
          break;

        case 35:
        case 'end':
          return _context3.stop();
      }
    }
  }, _marked[2], this, [[16, 20, 24, 32], [25,, 27, 31]]);
}
