// deep-iterator@1.1.0/lib/seen.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

exports.default = makeSeen;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var NoCheckSeen = function () {
  function NoCheckSeen() {
    (0, _classCallCheck3.default)(this, NoCheckSeen);
  }

  (0, _createClass3.default)(NoCheckSeen, [{
    key: 'add',
    value: function add() {
      return this;
    }
  }, {
    key: 'has',
    value: function has() {
      return false;
    }
  }]);
  return NoCheckSeen;
}();

// TODO : use an immutable collection


var LeafSeen = function () {
  function LeafSeen(arr) {
    (0, _classCallCheck3.default)(this, LeafSeen);

    this._arr = arr;
  }

  (0, _createClass3.default)(LeafSeen, [{
    key: 'add',
    value: function add(value) {
      return new LeafSeen([].concat((0, _toConsumableArray3.default)(this._arr), [value]));
    }
  }, {
    key: 'has',
    value: function has(value) {
      return this._arr.includes(value);
    }
  }]);
  return LeafSeen;
}();

// TODO : use an immutable collection


var ThrowSeen = function () {
  function ThrowSeen(arr) {
    (0, _classCallCheck3.default)(this, ThrowSeen);

    this._arr = arr;
  }

  (0, _createClass3.default)(ThrowSeen, [{
    key: 'add',
    value: function add(value) {
      return new ThrowSeen([].concat((0, _toConsumableArray3.default)(this._arr), [value]));
    }
  }, {
    key: 'has',
    value: function has(value) {
      if (this._arr.includes(value)) {
        throw new Error('Circular reference : ' + value);
      }
      return false;
    }
  }]);
  return ThrowSeen;
}();

function makeSeen(circularReference) {
  switch (circularReference) {
    case 'leaf':
      return new LeafSeen([]);
    case 'throw':
      return new ThrowSeen([]);
    case 'noCheck':
      return new NoCheckSeen();
  }
  throw new Error('Incorrect value ' + circularReference + ' for circularReference option.');
}
