// @saeris/graphql-directives@0.5.0/dist-src/currencies.js
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import { SchemaDirectiveVisitor } from "apollo-server";
import { defaultFieldResolver, GraphQLDirective, DirectiveLocation, GraphQLString, GraphQLEnumType } from "graphql";
import dinero from "dinero.js";
import { directiveToString, directiveToDocumentNode } from "./utils.js";
const RoundingMode = new GraphQLEnumType({
  name: `RoundingMode`,
  values: {
    HALF_ODD: {},
    HALF_EVEN: {},
    HALF_UP: {},
    HALF_DOWN: {},
    HALF_TOWARD_ZERO: {},
    HALF_AWAY_FROM_ZERO: {}
  }
});
export class formatCurrency extends SchemaDirectiveVisitor {
  static getDirectiveDeclaration() {
    return new GraphQLDirective({
      name: `formatCurrency`,
      locations: [DirectiveLocation.FIELD_DEFINITION],
      args: {
        defaultFormat: {
          type: GraphQLString,
          defaultValue: `$0,0.00`
        },
        defaultRoundingMode: {
          type: RoundingMode,
          defaultValue: `HALF_AWAY_FROM_ZERO`
        }
      }
    });
  }

  static toString() {
    return directiveToString(this.getDirectiveDeclaration());
  }

  static toDocumentNode() {
    return directiveToDocumentNode(this.getDirectiveDeclaration());
  }

  visitFieldDefinition(field) {
    const {
      resolve = defaultFieldResolver
    } = field;
    const {
      defaultFormat,
      defaultRoundingMode
    } = this.args;
    field.args.push({
      name: `format`,
      type: GraphQLString
    });
    field.args.push({
      name: `currency`,
      type: GraphQLString
    });
    field.args.push({
      name: `roundingMode`,
      type: RoundingMode
    });

    field.resolve = async function (source, _ref, context, info) {
      let {
        format,
        currency,
        roundingMode
      } = _ref,
          args = _objectWithoutProperties(_ref, ["format", "currency", "roundingMode"]);

      const result = await resolve.call(this, source, args, context, info);

      const transform = input => {
        const config = {
          amount: input
        };
        if (currency) config.currency = currency;
        return dinero(config).toFormat(format || defaultFormat, roundingMode || defaultRoundingMode);
      };

      return Array.isArray(result) ? result.map(transform) : transform(result);
    };

    field.type = GraphQLString;
  }

}
